core/extension/gdextension_interface.h

```diff
@@ -155,27 +155,45 @@ typedef enum {
 // - Some types have no destructor (see `extension_api.json`'s `has_destructor` field), for
 //   them it is always safe to skip the constructor for the return value if you are in a hurry ;-)
 
-typedef void *GDExtensionVariantPtr;
-typedef const void *GDExtensionConstVariantPtr;
-typedef void *GDExtensionUninitializedVariantPtr;
-typedef void *GDExtensionStringNamePtr;
-typedef const void *GDExtensionConstStringNamePtr;
-typedef void *GDExtensionUninitializedStringNamePtr;
-typedef void *GDExtensionStringPtr;
-typedef const void *GDExtensionConstStringPtr;
-typedef void *GDExtensionUninitializedStringPtr;
-typedef void *GDExtensionObjectPtr;
-typedef const void *GDExtensionConstObjectPtr;
-typedef void *GDExtensionUninitializedObjectPtr;
-typedef void *GDExtensionTypePtr;
-typedef const void *GDExtensionConstTypePtr;
-typedef void *GDExtensionUninitializedTypePtr;
-typedef const void *GDExtensionMethodBindPtr;
+#define GDEXTENSION_OPAQUE_PTR(type) \
+	typedef struct GDExtensionOpaque##type *GDExtension##type##Ptr;
+
+#define GDEXTENSION_CONST_OPAQUE_PTR(type) \
+	typedef const struct GDExtensionOpaque##type *GDExtensionConst##type##Ptr;
+
+#define GDEXTENSION_UNINITIALIZED_OPAQUE_PTR(type) \
+	typedef struct GDExtensionUninitialized##type *GDExtensionUninitialized##type##Ptr;
+
+GDEXTENSION_OPAQUE_PTR(Variant)
+GDEXTENSION_CONST_OPAQUE_PTR(Variant)
+GDEXTENSION_UNINITIALIZED_OPAQUE_PTR(Variant)
```

Comment:
I'm not sure about using preprocessor macros in this file.

For any binding that will compile C/C++ code, this is fine. But I wonder what effect this will have on bindings that are processing the `gdextension_interface.h` with something other than a C/C++ compiler to generate glue code? I think the Rust binding might work this way?

@Bromeon Your thoughts?

Reply:
Fully agree, please don't use macros here.

Even if repetitive, the code must be written only once, but it will be read many times. Having an extra abstraction here makes it more difficult to understand the types. It also gets in the way of automated tooling like binding generators.